go/store/datas/pull.go

```diff
@@ -102,56 +115,108 @@ func clone(ctx context.Context, srcTS, sinkTS nbs.TableFileStore, eventCh chan<-
 		return err
 	}
 
+	desiredFiles, fileIDToTF := mapTableFiles(tblFiles)
+
 	if eventCh != nil {
 		eventCh <- TableFileEvent{Listed, tblFiles}
 	}
 
-	// should parallelize at some point
-	for _, tblFile := range tblFiles {
-		if eventCh != nil {
-			eventCh <- TableFileEvent{DownloadStart, []nbs.TableFile{tblFile}}
-		}
+	i := 0
+	op := func() error {
+		var err error
+		for i < len(desiredFiles) {
+			fileID := desiredFiles[i]
+			tblFile, ok := fileIDToTF[fileID]
 
-		err := func() (err error) {
-			var rd io.ReadCloser
-			rd, err = tblFile.Open()
+			if !ok {
+				// conjoin happened during clone
+				return backoff.Permanent(errors.New("table file not found. please try again"))
+			}
 
-			if err != nil {
-				if eventCh != nil {
-					eventCh <- TableFileEvent{DownloadFailed, []nbs.TableFile{tblFile}}
+			err = func() (err error) {
+				var rd io.ReadCloser
+				rd, err = tblFile.Open()
+
+				if err != nil {
+					return err
 				}
 
-				return err
-			}
+				defer func() {
+					closeErr := rd.Close()
 
-			defer func() {
-				closeErr := rd.Close()
+					if err == nil && closeErr != nil {
+						err = closeErr
+					}
+				}()
 
-				if err == nil {
-					err = closeErr
+				if eventCh != nil {
+					eventCh <- TableFileEvent{DownloadStart, []nbs.TableFile{tblFile}}
 				}
-			}()
 
-			err = sinkTS.WriteTableFile(ctx, tblFile.FileID(), tblFile.NumChunks(), rd, 0, nil)
+				err = sinkTS.WriteTableFile(ctx, tblFile.FileID(), tblFile.NumChunks(), rd, 0, nil)
+
+				if err != nil {
+					if eventCh != nil {
+						eventCh <- TableFileEvent{DownloadFailed, []nbs.TableFile{tblFile}}
+					}
+
+					return err
+				}
 
-			if err != nil {
 				if eventCh != nil {
-					eventCh <- TableFileEvent{DownloadFailed, []nbs.TableFile{tblFile}}
+					eventCh <- TableFileEvent{DownloadSuccess, []nbs.TableFile{tblFile}}
 				}
 
-				return err
-			}
+				return nil
+			}()
 
-			if eventCh != nil {
-				eventCh <- TableFileEvent{DownloadSuccess, []nbs.TableFile{tblFile}}
+			if err != nil {
+				break
 			}
 
-			return nil
-		}()
+			i++
+		}
 
 		if err != nil {
+			var sourcesErr error
+			_, tblFiles, sourcesErr = srcTS.Sources(ctx)
```

Comment:
Not obvious why this would return a different set of table files than it did outside this function. Could use an explanatory comment

Reply:
It shouldn't return a different set.  But it will refresh the signed urls.  Not the best solution but doesn't require a server deployment.

Reply:
comments added.